{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In this file there are 25 records\n",
      "The lengths of the sequences in this file are: [990, 724, 3080, 2863, 3832, 4805, 1663, 512, 691, 3072, 1801, 3603, 2478, 1608, 4745, 1810, 3424, 1451, 3276, 2124, 1712, 1325, 1189, 555, 2449]\n",
      "The name of the longest sequence in this file is: gi|142022655|gb|EQ086233.1|323 with a length of 4805\n",
      "The name of the shortest sequence in this file is: gi|142022655|gb|EQ086233.1|521 with a length of 512\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence name</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Protein sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gi|142022655|gb|EQ086233.1|43</td>\n",
       "      <td>ATGGCCTCCAGCGCACCGATCGGATCAAAGCCGCTGAAGCCTTCGC...</td>\n",
       "      <td>MASSAPIGSKPLKPSRIRRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gi|142022655|gb|EQ086233.1|43</td>\n",
       "      <td>ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG</td>\n",
       "      <td>MRRSVIAALIE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gi|142022655|gb|EQ086233.1|43</td>\n",
       "      <td>ATGCCCAGCACGCCAATGCGTTCTTCATCCACA</td>\n",
       "      <td>MPSTPMRSSST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gi|142022655|gb|EQ086233.1|43</td>\n",
       "      <td>ATGCGTTCTTCATCCACA</td>\n",
       "      <td>MRSSST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gi|142022655|gb|EQ086233.1|43</td>\n",
       "      <td>ATGACGACGAAACCTTCCTTGGCCAGCGCCTCGCCATACACGTTCC...</td>\n",
       "      <td>MTTKPSLASASPYTFPDVCSLQLPIGCALMMAGYFLPSSKFGGKWM...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Sequence name  \\\n",
       "0  gi|142022655|gb|EQ086233.1|43   \n",
       "1  gi|142022655|gb|EQ086233.1|43   \n",
       "2  gi|142022655|gb|EQ086233.1|43   \n",
       "3  gi|142022655|gb|EQ086233.1|43   \n",
       "4  gi|142022655|gb|EQ086233.1|43   \n",
       "\n",
       "                                                Gene  \\\n",
       "0  ATGGCCTCCAGCGCACCGATCGGATCAAAGCCGCTGAAGCCTTCGC...   \n",
       "1                  ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG   \n",
       "2                  ATGCCCAGCACGCCAATGCGTTCTTCATCCACA   \n",
       "3                                 ATGCGTTCTTCATCCACA   \n",
       "4  ATGACGACGAAACCTTCCTTGGCCAGCGCCTCGCCATACACGTTCC...   \n",
       "\n",
       "                                    Protein sequence  \n",
       "0                               MASSAPIGSKPLKPSRIRRP  \n",
       "1                                        MRRSVIAALIE  \n",
       "2                                        MPSTPMRSSST  \n",
       "3                                             MRSSST  \n",
       "4  MTTKPSLASASPYTFPDVCSLQLPIGCALMMAGYFLPSSKFGGKWM...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "import pandas as pd\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "\n",
    "\n",
    "with open(\"/home/marwene/Downloads/dna.example.fasta\") as file:\n",
    "    sequences = {}\n",
    "    for line in file:\n",
    "        line = line.strip(\"\\n\")\n",
    "        if line.startswith(\">\"): # this is a name\n",
    "            words = line.split()\n",
    "            name = words[0][1:]\n",
    "            sequences[name] = \"\"\n",
    "        else: # this is a sequence\n",
    "            sequences[name] = sequences[name] + line\n",
    "\n",
    "# number of records in fasta file\n",
    "print(\"In this file there are {} records\".format(len(sequences.keys())))\n",
    "lengths_sequences = []\n",
    "for key, value in sequences.items():\n",
    "    lengths_sequences.append(len(value))\n",
    "\n",
    "# length of the sequences in a fasta file\n",
    "print(\"The lengths of the sequences in this file are: {}\".format(lengths_sequences))\n",
    "\n",
    "# name and length of the longest sequence\n",
    "for key, value in sequences.items():\n",
    "    if len(value) == max(lengths_sequences):\n",
    "        print(\"The name of the longest sequence in this file is: {} with a length of {}\".format(key, len(value)))\n",
    "\n",
    "\n",
    "# name and length of the shortest sequence\n",
    "for key, value in sequences.items():\n",
    "    if len(value) == min(lengths_sequences):\n",
    "        print(\"The name of the shortest sequence in this file is: {} with a length of {}\".format(key, len(value)))\n",
    "\n",
    "\n",
    "# find the coordinates of all the ORF in a fasta file\n",
    "ORF_coor = []\n",
    "for name, sequence in sequences.items():\n",
    "    for start_position in range(0, len(sequence) - 3):\n",
    "        codon = sequence[start_position: start_position + 3]\n",
    "        if codon == \"ATG\":\n",
    "            for stop_position in range(start_position, len(sequence) - 3, 3):\n",
    "                    codon = sequence[stop_position: stop_position + 3]\n",
    "                    if codon in [\"TAA\", \"TAG\", \"TGA\"]:\n",
    "                        ORF_coor.append((start_position, stop_position))\n",
    "                        break\n",
    "\n",
    "# Find all the ORF and append them to a list and then to a dictionnary with the key is the sequence \n",
    "#identifier and the value is the ORF sequence\n",
    "ORF_dict = {}\n",
    "ORF_seq = []\n",
    "\n",
    "\n",
    "def GEN():\n",
    "    for i in range(0, len(ORF_coor) - 1):\n",
    "        yield i\n",
    "\n",
    "\n",
    "a = GEN()\n",
    "\n",
    "for name, sequence in sequences.items():\n",
    "    for i in a:\n",
    "        if ORF_coor[i][0] > ORF_coor[i + 1][0]:\n",
    "            ORF_seq.append(sequence[ORF_coor[i][0]: ORF_coor[i][1]])\n",
    "            ORF_dict[name] = ORF_seq\n",
    "            ORF_seq = []\n",
    "            break\n",
    "        else:\n",
    "            ORF_seq.append(sequence[ORF_coor[i][0]: ORF_coor[i][1]])\n",
    "\n",
    "# This list contains the all the identifiers in a fasta file             \n",
    "seq_name = []\n",
    "# This list contains all the ORF translated using the Biopython \n",
    "ORF_translated = []\n",
    "# This list contains all the ORFs sequences\n",
    "ORF_gene = []\n",
    "\n",
    "for name, ORF_list in ORF_dict.items():\n",
    "    for ORF in ORF_list:\n",
    "        seq_name.append(name)\n",
    "        ORF_translated.append(Seq(ORF).translate())\n",
    "        ORF_gene.append(ORF)\n",
    "\n",
    "# create a data frame that contains name of the sequence, the sequence of the gene and the protein translated\n",
    "ORF_data_frame = pd.DataFrame()\n",
    "ORF_data_frame[\"Sequence name\"] = seq_name\n",
    "ORF_data_frame[\"Gene\"] = ORF_gene\n",
    "ORF_data_frame[\"Protein sequence\"] = [''.join(aa) for aa in ORF_translated] \n",
    "\n",
    "ORF_data_frame.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# choose one example of e gene\n",
    "gene = ORF_data_frame[\"Gene\"][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", Seq(gene))\n",
    "blast_record = NCBIXML.read(result_handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "****Alignment****\n",
      "sequence: gi|1784293010|gb|CP046607.1| Burkholderia contaminans strain XL73 chromosome 1, complete sequence\n",
      "length: 1541420\n",
      "e value: 5.04371e-07\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "|||||||||||||||||||||||||||||||||...\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "****Alignment****\n",
      "sequence: gi|1767888081|gb|CP028809.1| Burkholderia contaminans strain SK875 chromosome SK875-3, complete sequence\n",
      "length: 1528467\n",
      "e value: 5.04371e-07\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "|||||||||||||||||||||||||||||||||...\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "****Alignment****\n",
      "sequence: gi|1712371596|gb|CP042166.1| Burkholderia contaminans strain ZCC chromosome 3, complete sequence\n",
      "length: 1536075\n",
      "e value: 5.04371e-07\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "|||||||||||||||||||||||||||||||||...\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "****Alignment****\n",
      "sequence: gi|1240697943|dbj|AP018359.1| Burkholderia contaminans CH-1 DNA, scaffold: scaffold03\n",
      "length: 1538363\n",
      "e value: 5.04371e-07\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "|||||||||||||||||||||||||||||||||...\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "****Alignment****\n",
      "sequence: gi|1062058563|gb|CP013392.1| Burkholderia contaminans strain FL-1-2-30-S1-D0 chromosome 3, complete sequence\n",
      "length: 1370749\n",
      "e value: 5.04371e-07\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "|||||||||||||||||||||||||||||||||...\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "****Alignment****\n",
      "sequence: gi|834770562|gb|CP009745.1| Burkholderia contaminans strain MS14 chromosome 3, complete sequence\n",
      "length: 1627712\n",
      "e value: 5.04371e-07\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "|||||||||||||||||||||||||||||||||...\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "****Alignment****\n",
      "sequence: gi|973224854|gb|CP013732.1| Burkholderia cepacia JBK9 chromosome 3, complete sequence\n",
      "length: 1290392\n",
      "e value: 7.48553e-05\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGA...\n",
      "|||||||||||||||||||| |||||||||||...\n",
      "ATGCGGCGCTCGGTCATCGCGGCATTGATCGA...\n",
      "****Alignment****\n",
      "sequence: gi|1069768633|gb|CP016444.1| Burkholderia stabilis strain ATCC BAA-67 chromosome 3, complete sequence\n",
      "length: 1322975\n",
      "e value: 0.000911924\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGA...\n",
      "||||||||||||||||| || |||||||||||...\n",
      "ATGCGGCGCTCGGTCATGGCGGCATTGATCGA...\n",
      "****Alignment****\n",
      "sequence: gi|77964193|gb|CP000150.1| Burkholderia lata strain 383 chromosome 3, complete sequence\n",
      "length: 1395069\n",
      "e value: 0.000911924\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGA...\n",
      "|||||||| ||||||||||| |||||||||||...\n",
      "ATGCGGCGTTCGGTCATCGCCGCATTGATCGA...\n",
      "****Alignment****\n",
      "sequence: gi|1061746170|gb|CP013451.1| Burkholderia cenocepacia strain MSMB384WGS chromosome 3, complete sequence\n",
      "length: 1121886\n",
      "e value: 0.00318293\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCG...\n",
      "|||||||||||||||||||| || |||||||...\n",
      "ATGCGGCGCTCGGTCATCGCCGCGTTGATCG...\n",
      "****Alignment****\n",
      "sequence: gi|1061566715|gb|CP013378.1| Burkholderia pseudomultivorans strain SUB-INT23-BP2 chromosome 1, complete sequence\n",
      "length: 4849929\n",
      "e value: 0.00318293\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCG...\n",
      "|||||||||||||||||||| || |||||||...\n",
      "ATGCGGCGCTCGGTCATCGCCGCGTTGATCG...\n",
      "****Alignment****\n",
      "sequence: gi|1115993426|emb|LT671862.1| Ochrobactrum anthropi genome assembly, chromosome: II\n",
      "length: 2037532\n",
      "e value: 0.0111095\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "||||||||||||||||| || || |||||||||...\n",
      "ATGCGGCGCTCGGTCATAGCAGCGTTGATCGAG...\n",
      "****Alignment****\n",
      "sequence: gi|1929917788|gb|CP064062.1| Ochrobactrum anthropi strain PBO chromosome 1, complete sequence\n",
      "length: 2097628\n",
      "e value: 0.0111095\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "||||||||||||||||| || || |||||||||...\n",
      "ATGCGGCGCTCGGTCATAGCAGCGTTGATCGAG...\n",
      "****Alignment****\n",
      "sequence: gi|1910762289|gb|CP061773.1| Ochrobactrum sp. MT180101 chromosome 2, complete sequence\n",
      "length: 2089912\n",
      "e value: 0.0111095\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "||||||||||||||||| || || |||||||||...\n",
      "ATGCGGCGCTCGGTCATAGCAGCGTTGATCGAG...\n",
      "****Alignment****\n",
      "sequence: gi|77968738|gb|CP000152.1| Burkholderia lata strain 383 chromosome 2, complete sequence\n",
      "length: 3587082\n",
      "e value: 0.0111095\n",
      "CGGCGCTCGGTCATCGCTGCATTGATCG...\n",
      "||||||||||||||||||||||||| ||...\n",
      "CGGCGCTCGGTCATCGCTGCATTGACCG...\n",
      "****Alignment****\n",
      "sequence: gi|72117119|gb|CP000090.1| Ralstonia eutropha JMP134 chromosome 1, complete sequence\n",
      "length: 3806533\n",
      "e value: 0.0111095\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGAG...\n",
      "||||||||||| ||||| || ||||||||||||...\n",
      "ATGCGGCGCTCTGTCATGGCGGCATTGATCGAG...\n",
      "****Alignment****\n",
      "sequence: gi|568237447|gb|CP007012.1| Pseudomonas sp. FGI182, complete genome\n",
      "length: 5891312\n",
      "e value: 0.038776\n",
      "ATGCGGCGCTCGGTCATCGCTGCATTGATCGA...\n",
      "||||| || ||||||||||| |||||||||||...\n",
      "ATGCGCCGTTCGGTCATCGCAGCATTGATCGA...\n"
     ]
    }
   ],
   "source": [
    "#  quick summary of all of the alignments greater than some threshold value.\n",
    "E_VALUE_THRESH = 0.04\n",
    "\n",
    "for alignment in blast_record.alignments:\n",
    "    for hsp in alignment.hsps:\n",
    "        if hsp.expect < E_VALUE_THRESH:\n",
    "            print('****Alignment****')\n",
    "            print('sequence:', alignment.title)\n",
    "            print('length:', alignment.length)\n",
    "            print('e value:', hsp.expect)\n",
    "            print(hsp.query[0:75] + '...')\n",
    "            print(hsp.match[0:75] + '...')\n",
    "            print(hsp.sbjct[0:75] + '...')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BIO1",
   "language": "python",
   "name": "bio1"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
